{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "assured-austria",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#################################################  Start #############################################################\n",
    "\n",
    "# At first, the DNA sequences and the Amino Acid Sequences are downloaded from NCBI website using code NM_207618.2\n",
    "# To download, search on Google as 'NCBI' >>> Click on NCBI official web link >>> \n",
    "# >>> select 'Nucleotide' and write 'NM_207618.2' on the search bar >>> search >> \n",
    "# >>> a web page would be opened titled as \"Mus musculus vomeronasal 1 receptor, D18 (V1rd18), mRNA\" \n",
    "# >>> click on 'FASTA' >>> copy it and save it\n",
    "# >>> go back at the NM_207618.2 web page and click on CDS >> copy it and save it\n",
    "# the DNA and the Amino Acid Sequences are saved in a two different txt files e.g. 'dna.txt' and 'protein.txt'\n",
    "\n",
    "# A function has been created to remove all special characters from our downloaded files\n",
    "\n",
    "def read_seq(inputfile):\n",
    "    \"\"\" Read and returns the input sequence with special characters removed \"\"\" \n",
    "    #The above script written within (\" \") is docstring. You could access it by help(read_seq)\n",
    "    with open(inputfile, 'r') as f:\n",
    "        seq=f.read()\n",
    "    seq = seq.replace('\\n','')\n",
    "    seq = seq.replace('\\r','')\n",
    "    return seq\n",
    "\n",
    "# Now the translation function is written\n",
    "\n",
    "def translate (seq):\n",
    "    \"\"\"\n",
    "    Translate a string containing a nucleotide sequence into a string containing the corresponding sequence of \n",
    "    amino acids . Nucleotides are translated in triplets using the table dictionary; each amino acid 4 is encoded \n",
    "    with a string of length 1.\n",
    "    \"\"\"\n",
    "    table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "    }\n",
    "    # check that the sequence length is divisible by 3\n",
    "    protein=''\n",
    "    if len(seq)%3 == 0:\n",
    "        # loop over the sequence\n",
    "        for i in range(0, len(seq), 3):\n",
    "            # extract a single codon\n",
    "            codon =seq[i : i+3]\n",
    "            # look up the codon and store the result\n",
    "            protein += table[codon]\n",
    "    return protein\n",
    "\n",
    "# check the translate function using a single code \n",
    "translate('ATA')\n",
    "# or\n",
    "translate('ATA') # works fine\n",
    "\n",
    "\n",
    "# Now, import our txt file in Python using our predefined read_seq function\n",
    "\n",
    "prt = read_seq('protein.txt')\n",
    "dna = read_seq('dna.txt')\n",
    "\n",
    "translate(dna) # gives you an empty string. This is because the length of dna is not divisible by 3. let's check\n",
    "len(dna)%3 #this gives the remainder 2. so length of dna is not divisible by 3. But why?\n",
    "# This is because, in NCBI website, we could see that it is written as \"CDS  21....938\" - this refers to the \n",
    "# locations of the gene where the coding sequence starts and ends\n",
    "\n",
    "# so for translating DNA, we would start at 21 and ends at 938. However, in Python index starts at 0 and ends \n",
    "# before the final index. we use dna[20:938]\n",
    "\n",
    "translate(dna[20:938])  #this gives you the translated amino acid code\n",
    "\n",
    "# now, check the NCBI official amino acid code\n",
    "prt\n",
    "\n",
    "# the translate(dna[20:938]) and prt are almost same except translate(dna[20:938]) prints an extra underscore(_). \n",
    "# This underscore(_) corresponds to stop codon which is ignored in NCBI's amino acid code. so, let's ignore it.\n",
    "\n",
    "translate(dna[20:935])\n",
    "\n",
    "# let's check whether our code matches to the official code\n",
    "\n",
    "prt == translate(dna[20:935])\n",
    "\n",
    "# or use the following code\n",
    "prt == translate(dna[20:938])[:-1]\n",
    "\n",
    "# it should print true which means our function works perfectly\n",
    "\n",
    "\n",
    "################################################# The end #############################################################"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
